File: Backtracking/306. Additive Number.cpp
// https://leetcode.com/problems/additive-number/

// same as: 842. Split Array into Fibonacci Sequence

class Solution {
	vector<string> cur;

public:
	bool isAdditiveNumber(string& s, int start = 0) {
		if (start == s.size() and cur.size() >= 3)
			return true;

		for (int i = start; i < s.size(); i++) {
			if (s[start] == '0' and i > start) break;
			string N = s.substr(start, i - start + 1);
			int l = cur.size();
			
			if (l <= 1 or addition(cur[l-1], cur[l-2]) == N) {
				cur.push_back(N);
				if (isAdditiveNumber(s, i+1))
					return true;
				cur.pop_back();
			}
		}

		return false;
	}

	string addition(string a, string b) {
		string ans;
		reverse(a.begin(), a.end());
		reverse(b.begin(), b.end());
		int carry = 0;
		for (int i = 0; i < a.size() or i < b.size(); i++) {
			int n1 = i < a.size() ? a[i] - '0' : 0;
			int n2 = i < b.size() ? b[i] - '0' : 0;
			ans.push_back('0' + (n1 + n2 + carry) % 10);
			carry = (n1 + n2 + carry) / 10 ;
		}
		if (carry > 0) ans.push_back('0' + carry);
		return string(ans.rbegin(), ans.rend());
	}
};

File: Backtracking/306. Additive Number.py
# https://leetcode.com/problems/additive-number/submissions/

class Solution:
	def isAdditiveNumber(self, s) -> bool:

		def backtrack(s, start) -> bool:	
			if start == len(s) and len(self.cur) >= 3: return True
			
			for i in range(start, len(s)):
				if s[start] == '0' and i > start: break
				N = int(s[start : i+1])
				l = len(self.cur)
				if l <= 1 or self.cur[l-1] + self.cur[l-2] == N:
					self.cur.append(N)
					if (backtrack(s, i+1)):
						return True
					self.cur.pop()

			return False

		self.cur = []
		return backtrack(s, 0)